src/common/utils/registry/auth/authorizer.go

```diff
@@ -0,0 +1,122 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"fmt"
+	"github.com/docker/distribution/registry/client/auth/challenge"
+	"github.com/goharbor/harbor/src/common/http/modifier"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// NewAuthorizer creates an authorizer that can handle different auth schemes
+func NewAuthorizer(credential Credential, client ...*http.Client) modifier.Modifier {
+	authorizer := &authorizer{
+		credential: credential,
+	}
+	if len(client) > 0 {
+		authorizer.client = client[0]
+	}
+	if authorizer.client == nil {
+		authorizer.client = http.DefaultClient
+	}
+
+	return authorizer
+}
+
+// authorizer authorizes the request with the provided credential.
+// It determines the auth scheme of registry automatically and calls
+// different underlying authorizers to do the auth work
+type authorizer struct {
+	client     *http.Client
+	url        *url.URL          // registry URL
+	authorizer modifier.Modifier // the underlying authorizer
+	credential Credential
+}
+
+func (a *authorizer) Modify(req *http.Request) error {
+	// Nil URL means this is the first time the authorizer is called
+	// Try to connect to the registry and determine the auth method
+	if a.url == nil {
+		if err := a.initialize(req.URL); err != nil {
+			return err
+		}
+	}
+
+	// check whether the request targets the registry
+	if !a.isTarget(req) {
+		return nil
+	}
+
+	return a.authorizer.Modify(req)
+}
+
+func (a *authorizer) initialize(u *url.URL) error {
+	url, err := url.Parse(u.Scheme + "://" + u.Host)
+	if err != nil {
+		return err
+	}
+	a.url = url
+	resp, err := a.client.Get(a.url.String() + "/v2/")
+	if err != nil {
+		return err
+	}
+	challenges := ParseChallengeFromResponse(resp)
+	// no challenge, mean no auth
+	if len(challenges) == 0 {
+		a.authorizer = &nullAuthorizer{}
+		return nil
+	}
+	cm := map[string]challenge.Challenge{}
+	for _, challenge := range challenges {
+		cm[challenge.Scheme] = challenge
+	}
+	if _, exist := cm["basic"]; exist {
+		a.authorizer = a.credential
```

Comment:
No underlying basic authorizer definition? We should follow a similar authorizer interface.

Reply:
This is tricky. The `credential` is an interface, it can be either basic auth authorier or secret authorizer. The secret is used in the replication  job to call registry APIs.

Reply:
`credential ` implements the authorizer interface? It's a little weird, why not new basic/secret authorized with passing the credential as inputs of `New()`？

Reply:
Opened issue to track the work https://github.com/goharbor/harbor/issues/10705